File: protocol-core/src/main/kotlin/com/airninja/protocol/pairing/Sas.kt
package com.airninja.protocol.pairing

import com.airninja.protocol.crypto.Sha256
import java.nio.ByteBuffer

object Sas {
    private const val PREFIX = "AIRNINJA-SAS"
    private const val MODULO = 1_000_000
    private const val DIGITS = 6
    private const val UNSIGNED_INT_MASK = 0xFFFFFFFFL

    fun derive(handshakeHash: ByteArray): String {
        val digest = Sha256.hash(PREFIX.toByteArray(Charsets.US_ASCII) + handshakeHash)
        val value = ByteBuffer.wrap(digest, 0, Int.SIZE_BYTES).int.toLong() and UNSIGNED_INT_MASK
        return (value % MODULO).toString().padStart(DIGITS, '0')
    }
}

File: macos/AirNinjaCore/Sources/AirNinjaCore/Sas.swift
import Foundation

public enum Sas {
    private static let prefix = "AIRNINJA-SAS"
    private static let modulo: UInt32 = 1_000_000

    public static func derive(handshakeHash: Data) -> String {
        var input = Data(prefix.utf8)
        input.append(handshakeHash)
        let digest = Sha256.hash(input)
        let value = (UInt32(digest[0]) << 24)
            | (UInt32(digest[1]) << 16)
            | (UInt32(digest[2]) << 8)
            | UInt32(digest[3])
        return String(format: "%06u", value % modulo)
    }
}
